Description: Fix a error about thread_local variables export to DLL, when compiled with clang
A related issue: https://github.com/drogonframework/drogon/issues/1845

From f63cf4953dfb495eaece459a3acc0895604537ce Mon Sep 17 00:00:00 2001
From: AgOH <stagoh17@gmail.com>
Date: Fri, 5 Apr 2024 05:29:24 +0800
Subject: [PATCH] Remove thread_local in dllexported function

---
 lib/src/Utilities.cc | 18 +++++++-----------
 1 file changed, 7 insertions(+), 11 deletions(-)

diff --git a/lib/src/Utilities.cc b/lib/src/Utilities.cc
index fff417d6ab..df2a1d57c2 100644
--- a/lib/src/Utilities.cc
+++ b/lib/src/Utilities.cc
@@ -1272,10 +1272,11 @@ std::string secureRandomString(size_t size)
 
     // batch up to 32 bytes of random data for efficiency. Calling
     // secureRandomBytes can be expensive.
-    auto randByte = []() {
-        thread_local trantor::utils::Hash256 hash;
-        thread_local size_t i = 0;
-        if (i == 0)
+    trantor::utils::Hash256 hash;
+    for (size_t i = 0; i < size; ++i)
+    {
+        auto j = i % sizeof(hash);
+        if (j == 0)
         {
             bool ok = trantor::utils::secureRandomBytes(&hash, sizeof(hash));
             if (!ok)
@@ -1283,13 +1284,8 @@ std::string secureRandomString(size_t size)
                     "Failed to generate random bytes for secureRandomString");
         }
         unsigned char *hashBytes = reinterpret_cast<unsigned char *>(&hash);
-        auto ret = hashBytes[i];
-        i = (i + 1) % sizeof(hash);
-        return ret;
-    };
-
-    for (size_t i = 0; i < size; ++i)
-        ret[i] = chars[randByte() % 64];
+        ret[i] = chars[hashBytes[j] % 64];
+    }
     return ret;
 }
 
